docs: update CHANGELOG for 0.13.0

cpu opened the pull request:
I've included substantially more detail in this CHANGELOG update compared to previous. To try and keep the top level content manageable, while also balancing making updates as friction-less as possible, I've split off some separate migration help to its own document.

I've done my best to flag all of the important changes but would appreciate more eyes to make sure I haven't missed or misrepresented anything. Like previous updates I've called out contributions by external folks and not those by core maintainers.

djc commented:
I would reorder these with breaking changes before added and with both lists ordered from most to least impact.

cpu commented:
> I would reorder these with breaking changes before added and with both lists ordered from most to least impact.

Good idea. I've done a reordering pass.

djc commented on rcgen/CHANGELOG.md:
```diff
@@ -1,6 +1,101 @@
+  `CertificateParams::signed_by`. You can serialize the `Certificate` to PEM or
+  DER using `Certificate::der()` or `Certificate::pem`.
+
+- `KeyPair` generation has been separted from certificate issuance. The
```
"separted"

cpu replied:
Fixed.

djc commented on rcgen/CHANGELOG.md:
```diff
@@ -1,6 +1,101 @@
+- `KeyPair` generation has been separted from certificate issuance. The
+  `Certificate::get_key_pair`, `Certificate::serialize_private_key_der` and
+  `Certificate::serialize_private_key_pem` fns have been removed. The
+  `Certificate` and `CertificateParams` types now only holds public data.
```
"holds" > "hold"

cpu replied:
Fixed.

djc commented on rcgen/CHANGELOG.md:
```diff
@@ -1,6 +1,101 @@
+  `KeyPair`, and the caller must generate their own `KeyPair` up-front using
+  `KeyPair::generate`, `KeyPair::generate_for`, or `KeyPair::generate_rsa_for`.
+
+- Certificate signing request creation and issuance has been reworked for
```
I think we need a similar section for CRL signing?

cpu replied:
Added coverage for CRLs.

djc commented on rcgen/CHANGELOG.md:
```diff
@@ -1,6 +1,101 @@
+  `Certificate`.
+
+- Throughout the API DER inputs are now represented using types from the Rustls
+  `pki-types` crate, e.g. `PrivateKeyDer`, `CertificateDer`,
```
I think we should call it by its crates.io name, rustls-pki-types.

cpu replied:
Fixed.

djc reviewed (commented):
Order looks great!

I think it's great to spend more words on the larger changes in the signing API, but (especially when also adding CRLs) I wonder if it could be done more concisely by explaining the general principles and abstracting as much as possible over certificate/CSR/CRL. I've tried hard to make the APIs as similar as possible for each of them.

I like to retain the `()` suffix on methods/functions because it makes it more obvious to distinguish these from types/variables.

cpu commented:
>  I wonder if it could be done more concisely by explaining the general principles and abstracting as much as possible over certificate/CSR/CRL.

Yeah, I think that's good feedback :+1: I will try and rework along these lines tomorrow.

cpu commented:
> I think it's great to spend more words on the larger changes in the signing API, but (especially when also adding CRLs) I wonder if it could be done more concisely by explaining the general principles and abstracting as much as possible over certificate/CSR/CRL. I've tried hard to make the APIs as similar as possible for each of them.

I went back and replaced the `CHANGELOG.md` coverage with a section aiming to explain the general principles without explicitly listing the mechanical changes. I think that's a good level of detail for a CHANGELOG and fits better with the existing practices.

But... I think the Rustls releases & feedback have shown me that there's an audience for material that is much more explicit about what changed and what should be used instead. I've split off some of that content into a separate `rcgen/docs/0.12-to-0.13.md` document. It's probably overkill to do that for every release but I felt like it was worthwhile for this one. I'd like to try and make sure the process is as painless as possible for consumers.

> I like to retain the () suffix on methods/functions because it makes it more obvious to distinguish these from types/variables.

Makes sense. I've done a pass adding them in both for the CHANGELOG and migration guide document.

cpu commented on rcgen/CHANGELOG.md:
```diff
@@ -1,6 +1,78 @@
+  `Certificate::get_times()` -> `Certificate::times()`
+
+- `KeyPair::from_pem_and_sign_algo()` has been renamed to
+  `KeyPair::from_pkcs8_pem_and_sign_algo()` to emphasize the input DER should be
```
Good call. I will catch up with the state of #242.

cpu replied:
Updated.

cpu commented on rcgen/docs/0.12-to-0.13.md:
```diff
@@ -0,0 +1,90 @@
+  the `cert` field of `CertifiedKey`, and the generated subject key pair in
+  `key_pair`.
+
+* Customized self-signed certificate issuance was previously done by
```
Fixed both.

cpu commented on rcgen/docs/0.12-to-0.13.md:
```diff
@@ -0,0 +1,90 @@
+  `Certificate::serialize_pem_with_signer()`. Each time a serialization fn was
+  called a new certificate was issued, leading to confusion when it was desired
+  to serialize the same certificate in two formats. In the new API issuance is
+  handled by `CertificateParams` fns and a `Certificate` is stable. You can
```
> or explicitly say "and the generated Certificate does not change"?

I went with a small tweak on this language saying "the generated `Certificate` will not change when serialized". Good suggestion, thanks.

djc replied:
It kind of still feels like a silly thing to say because the `der()` method doesn't even really do serialization and the new API just intrinsically makes it impossible for the previous pitfall to happen... not really sure how best to explain this.

cpu replied:
> It kind of still feels like a silly thing to say because the der() method doesn't even really do serialization and the new API just intrinsically makes it impossible for the previous pitfall to happen.

I think the reason it feels silly is because the old API was surprising and the new behaviour seems intuitively like what most folks would expect. 

I still think the text is better explicitly saying both what the old API did and what the new API doesn't do. I suspect many readers won't have realized what the old API was doing until this moment and might want to check their own usage for bugs.

> the der() method doesn't even really do serialization

No argument there. I changed the last sentence to say "You can serialize it to PEM by calling `Certificate::pem()`, or access the DER encoding by calling `Certificate::der()`.

cpu commented:
@est31 I think this is ready for review now and the last thing blocking a 0.13 release. If you're comfortable approving the branch can I go ahead and publish the release afterwards? It would be great if we could do this before you're availability is reduced :-)

est31 reviewed (approved):
LGTM, let's release 0.13

djc commented:
* Created `v0.13.0` tag
* Published rcgen v0.13.0 at registry `crates-io`

cpu commented:
Thanks for finishing this up djc :rocket:
